### PR TITLE
CP-1835 Pass Arguments to Local Tasks

### PR DESCRIPTION
**Pull request by bencampbell-wf**
Problem
-------

Any arguments used when calling dart_dev are parsed by the dart dev argument parser. It is not possible for locally discovered tasks to register arguments preventing them from being parsed as valid arguments. This prevents users from passing any arguments to the underlying local task.

Solution
--------

Local tasks can now take a list of unparsed arguments as an optional argument. If provided this list is passed to the invocation process allowing the task to obtain a list of all parameters after the local
task name.

In addtion, when dart dev attempts to parse the arguments for the first time, the list is scanned for the first commmand matching argument. It is checked to confirm it is a local task. The list of arguments leading up to the command are parsed allowing dart_dev global flags to be used
with local tasks.

Bonus
-----

A flag for expanding on symlinks within task search paths is provided in the local configuration. This allows consumers to limit the search for tasks to unlinked directories. This prevents dart dev from searching the `packages` symlink directory. 

Howto Test
----------

- [ ] Add this branch as a dependency override in a project that has a local task
- [ ] All dart_dev native tasks work as expected
- [ ] The local task should work without any extra arguments
- [ ] The local task should work with additional arguments provided after the local command name
- [ ] The local task should allow dart_dev global parameters (ie: --quite)

References
----------

- @evanweible-wf @jayudey-wf @maxwellpeterson-wf @trentgrover-wf

### CODE REVIEW

**Comment by evanweible-wf** [comment]
+1

**Comment by trentgrover-wf** [comment]
+1
@jayudey-wf ready for merge

**Comment by jayudey-wf** [comment]
talked with ben, he's adding an integration test to cover the new functionality

**Comment by bencampbell-wf** [comment]
@jayudey-wf I've added integration tests to confirm local tasks will receive all types of parameters. I've also added integration tests for symlink directory following.

File: test_fixtures/coverage/failing_test/test/passing_test.dart.temp.html [hunk 1]
```diff
@@ -0,0 +1 @@
+<script type="application/dart" src="passing_test.dart"></script>
```
Comment (jayudey-wf) [inline test_fixtures/coverage/failing_test/test/passing_test.dart.temp.html]: I don't think these files should be committed

**Comment by evanweible-wf** [comment]
+1

**Comment by bencampbell-wf** [comment]
@jayudey-wf this is ready for your quality review.

File: README.md [hunk 2]
```diff
@@ -93,9 +93,17 @@ local tasks.
...
-program. 
+Any file matching this pattern is parsed and registered as a task.  The task
+discovery will not follow symlink directories by default. Symlink directory
+expansion can be enabled using the `followSymlinks` property in the `local`
+configuration field.
+
+Any additional arguments specified after the command name are passed to the
+underlying task process as additional command line arguments. For instance `pub
+run dart_dev exampleTask --example-argument 23` would pass the example argument
+and value as additions to call to `dart exampleTask_task.dart
```
Comment (trentgrover-wf) [inline README.md]: `to the call to`?

File: README.md [hunk 3]
```diff
@@ -93,9 +93,17 @@ local tasks.
...
+--example-argument 23`. This allows the task to parse it's arguments
```
Comment (trentgrover-wf) [inline README.md]: `it's` -> `its`

**Comment by bencampbell-wf** [comment]
The latest stable release has corrected the failing tests for the coverage command!

**Comment by trentgrover-wf** [comment]
+1

**Comment by evanweible-wf** [comment]
+1

**Comment by jayudey-wf** [comment]
QA Resource Approval: +10

- [ ] Testing instruction
- [ ] Dev +1's
- [ ] Dev/QA +10 with detail of what was tested
 - installed locally and verified that general dart_dev tasks were working as expected, local tasks were working as expected, that arguments could be passed to local tasks, that global flags could be passed to local tasks and that sym links were being search as expected
- [ ] Unit test created/updated
- [ ] All unit tests pass

Merging into master.

**Pull request merged**